<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>XGR+ User Manual</title>
  <meta name="description" content="XGR+ User Manual" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="XGR+ User Manual" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="XGR+ User Manual" />
  
  
  



<meta name="date" content="2023-03-11" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="landing-frontpage.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>




</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">XGR+ User Manual</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> OVERVIEW</a></li>
<li class="chapter" data-level="2" data-path="landing-frontpage.html"><a href="landing-frontpage.html"><i class="fa fa-check"></i><b>2</b> LANDING FRONTPAGE</a></li>
<li class="chapter" data-level="3" data-path="mobile-first-webserver.html"><a href="mobile-first-webserver.html"><i class="fa fa-check"></i><b>3</b> MOBILE-FIRST WEBSERVER</a></li>
<li class="chapter" data-level="4" data-path="compatibility-runtime.html"><a href="compatibility-runtime.html"><i class="fa fa-check"></i><b>4</b> COMPATIBILITY &amp; RUNTIME</a></li>
<li class="chapter" data-level="5" data-path="help-buttons.html"><a href="help-buttons.html"><i class="fa fa-check"></i><b>5</b> HELP BUTTONS</a></li>
<li class="chapter" data-level="6" data-path="error-messages.html"><a href="error-messages.html"><i class="fa fa-check"></i><b>6</b> ERROR MESSAGES</a></li>
<li class="chapter" data-level="7" data-path="enrichment-analyser-genes---eag.html"><a href="enrichment-analyser-genes---eag.html"><i class="fa fa-check"></i><b>7</b> ENRICHMENT ANALYSER (GENES) - EAG</a>
<ul>
<li class="chapter" data-level="7.1" data-path="enrichment-analyser-genes---eag.html"><a href="enrichment-analyser-genes---eag.html#input-interface"><i class="fa fa-check"></i><b>7.1</b> Input Interface</a></li>
<li class="chapter" data-level="7.2" data-path="enrichment-analyser-genes---eag.html"><a href="enrichment-analyser-genes---eag.html#enrichment-results"><i class="fa fa-check"></i><b>7.2</b> Enrichment Results</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="enrichment-analyser-snps---eas.html"><a href="enrichment-analyser-snps---eas.html"><i class="fa fa-check"></i><b>8</b> ENRICHMENT ANALYSER (SNPS) - EAS</a>
<ul>
<li class="chapter" data-level="8.1" data-path="enrichment-analyser-snps---eas.html"><a href="enrichment-analyser-snps---eas.html#input-interface-1"><i class="fa fa-check"></i><b>8.1</b> Input Interface</a></li>
<li class="chapter" data-level="8.2" data-path="enrichment-analyser-snps---eas.html"><a href="enrichment-analyser-snps---eas.html#enrichment-results-1"><i class="fa fa-check"></i><b>8.2</b> Enrichment Results</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="enrichment-analyser-regions---ear.html"><a href="enrichment-analyser-regions---ear.html"><i class="fa fa-check"></i><b>9</b> ENRICHMENT ANALYSER (REGIONS) - EAR</a>
<ul>
<li class="chapter" data-level="9.1" data-path="enrichment-analyser-regions---ear.html"><a href="enrichment-analyser-regions---ear.html#input-interface-2"><i class="fa fa-check"></i><b>9.1</b> Input Interface</a></li>
<li class="chapter" data-level="9.2" data-path="enrichment-analyser-regions---ear.html"><a href="enrichment-analyser-regions---ear.html#enrichment-results-2"><i class="fa fa-check"></i><b>9.2</b> Enrichment Results</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="enrichment-analyser-domains---ead.html"><a href="enrichment-analyser-domains---ead.html"><i class="fa fa-check"></i><b>10</b> ENRICHMENT ANALYSER (DOMAINS) - EAD</a>
<ul>
<li class="chapter" data-level="10.1" data-path="enrichment-analyser-domains---ead.html"><a href="enrichment-analyser-domains---ead.html#input-interface-3"><i class="fa fa-check"></i><b>10.1</b> Input Interface</a></li>
<li class="chapter" data-level="10.2" data-path="enrichment-analyser-domains---ead.html"><a href="enrichment-analyser-domains---ead.html#enrichment-results-3"><i class="fa fa-check"></i><b>10.2</b> Enrichment Results</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="subnetwork-analyser-genes---sag.html"><a href="subnetwork-analyser-genes---sag.html"><i class="fa fa-check"></i><b>11</b> SUBNETWORK ANALYSER (GENES) - SAG</a>
<ul>
<li class="chapter" data-level="11.1" data-path="subnetwork-analyser-genes---sag.html"><a href="subnetwork-analyser-genes---sag.html#input-interface-4"><i class="fa fa-check"></i><b>11.1</b> Input Interface</a></li>
<li class="chapter" data-level="11.2" data-path="subnetwork-analyser-genes---sag.html"><a href="subnetwork-analyser-genes---sag.html#subnetwork-results"><i class="fa fa-check"></i><b>11.2</b> Subnetwork Results</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="subnetwork-analyser-snps---sas.html"><a href="subnetwork-analyser-snps---sas.html"><i class="fa fa-check"></i><b>12</b> SUBNETWORK ANALYSER (SNPS) - SAS</a>
<ul>
<li class="chapter" data-level="12.1" data-path="subnetwork-analyser-snps---sas.html"><a href="subnetwork-analyser-snps---sas.html#input-interface-5"><i class="fa fa-check"></i><b>12.1</b> Input Interface</a></li>
<li class="chapter" data-level="12.2" data-path="subnetwork-analyser-snps---sas.html"><a href="subnetwork-analyser-snps---sas.html#subnetwork-results-1"><i class="fa fa-check"></i><b>12.2</b> Subnetwork Results</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="subnetwork-analyser-regions---sar.html"><a href="subnetwork-analyser-regions---sar.html"><i class="fa fa-check"></i><b>13</b> SUBNETWORK ANALYSER (REGIONS) - SAR</a>
<ul>
<li class="chapter" data-level="13.1" data-path="subnetwork-analyser-regions---sar.html"><a href="subnetwork-analyser-regions---sar.html#input-interface-6"><i class="fa fa-check"></i><b>13.1</b> Input Interface</a></li>
<li class="chapter" data-level="13.2" data-path="subnetwork-analyser-regions---sar.html"><a href="subnetwork-analyser-regions---sar.html#subnetwork-results-2"><i class="fa fa-check"></i><b>13.2</b> Subnetwork Results</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="/XGRplus" target="blank">XGR+</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">XGR+ User Manual</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">XGR+ User Manual</h1>
<p class="date"><em>2023-03-11</em></p>
</div>
<div id="index" class="section level1 hasAnchor" number="1">
<h1><span class="header-section-number">Section 1</span> OVERVIEW<a href="index.html#index" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:logo"></span>
<img src="index_files/figure-html/logo-1.png" alt="The logo for eXploring Genomic Relations + (XGR PLUS)." width="25%" />
<p class="caption">
FIGURE 1.1: The logo for eXploring Genomic Relations + (XGR PLUS).
</p>
</div>
<hr>
<blockquote>
<p><strong>Motivation</strong></p>
</blockquote>
<p>We are very pleased to introduce an updated version of our XGR software, which has gained increasing popularity in genomic summary data interpretation since its first release (see <a href="https://www.ncbi.nlm.nih.gov/pubmed/27964755">Genome Medicine 2016</a>). This update features a newly designed web server, called XGR PLUS or XGR+, now made available <a href="http://www.genomicsummary.pro/XGRplus">here</a> to all users for free and without any login requirements. It enables almost real-time enrichment and subnetwork analyses for a user-input list of genes, SNPs, genomic regions, or protein domains, leveraging ontologies, networks, and knowledgebase resources about functional genomic datasets, such as e/pQTL, promoter capture Hi-C, and enhancer-gene maps. Besides, the web server is user-friendly, very suitable for users who are unfamiliar with coding.</p>
<p>One of the major challenges in genomics research is how to bridge the gap between generated genomic summary data and downstream knowledge discovery. We define genomic summary data as a list of genes or SNPs (more generally, genomic regions), along with their significance levels (e.g.Â p-values). We are leading the field in making sense of genomic summary data for knowledge discovery through developing approaches and tools that have been verified to be efficient and effective, including XGR (see <a href="https://www.ncbi.nlm.nih.gov/pubmed/27964755">Genome Medicine 2016</a>), dnet (see <a href="https://www.ncbi.nlm.nih.gov/pubmed/25246945">Genome Medicine 2014</a>), Priority index (see <a href="https://www.ncbi.nlm.nih.gov/pubmed/31253980">Nature Genetics 2019</a>), and dcGO (see <a href="https://www.ncbi.nlm.nih.gov/pubmed/23161684">Nucleic Acids Research 2013</a>).</p>
<p>Building upon our previous approaches that have stood the test of time (with most of citations received in the past two years), and also incorporating the growing knowledgebase of ontologies and networks, XGR+ offers seven analysers for enrichment and subnetwork analyses, each doing specific interpretations tailored to genomic summary data at the gene, SNP, genomic region, and protein domain levels.</p>
<blockquote>
<p><strong>Enrichment Analysers</strong></p>
</blockquote>
<ul>
<li><p><a href="enrichment-analyser-genes---eag.html#enrichment-analyser-genes---eag">ENRICHMENT ANALYSER (GENES) - EAG</a>, identifies ontology terms enriched for input genes; see <a href="/app/examples/_tmp_RMD_EAgene.html">Example Output</a></p></li>
<li><p><a href="enrichment-analyser-snps---eas.html#enrichment-analyser-snps---eas">ENRICHMENT ANALYSER (SNPS) - EAS</a>, identifies ontology terms enriched for genes linked from input SNPs; see <a href="/app/examples/_tmp_RMD_EAsnp.html">Example Output</a></p></li>
<li><p><a href="enrichment-analyser-regions---ear.html#enrichment-analyser-regions---ear">ENRICHMENT ANALYSER (REGIONS) - EAR</a>, identifies ontology terms enriched for genes linked from input genomic regions; see <a href="/app/examples/_tmp_RMD_EAregion.html">Example Output</a></p></li>
<li><p><a href="enrichment-analyser-domains---ead.html#enrichment-analyser-domains---ead">ENRICHMENT ANALYSER (DOMAINS) - EAD</a>, identifies ontology terms enriched for input protein domains; see <a href="/app/examples/_tmp_RMD_EAdomain.html">Example Output</a></p></li>
</ul>
<blockquote>
<p><strong>Subnetwork Analysers</strong></p>
</blockquote>
<ul>
<li><p><a href="subnetwork-analyser-genes---sag.html#subnetwork-analyser-genes---sag">SUBNETWORK ANALYSER (GENES) - SAG</a>, identifies a gene subnetwork based on input gene-level summary data; see <a href="/app/examples/_tmp_RMD_SAgene.html">Example Output</a></p></li>
<li><p><a href="subnetwork-analyser-snps---sas.html#subnetwork-analyser-snps---sas">SUBNETWORK ANALYSER (SNPS) - SAS</a>, identifies a gene subnetwork based on genes linked from input SNP-level summary data; see <a href="/app/examples/_tmp_RMD_SAsnp.html">Example Output</a></p></li>
<li><p><a href="subnetwork-analyser-regions---sar.html#subnetwork-analyser-regions---sar">SUBNETWORK ANALYSER (REGIONS) - SAR</a>, identifies a gene subnetwork based on genes linked from input genomic region-level summary data; see <a href="/app/examples/_tmp_RMD_SAregion.html">Example Output</a></p></li>
</ul>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="landing-frontpage.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "serif",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "lunr",
"options": null
},
"toc": {
"collapse": "subsection",
"scroll_highlight": true
},
"toolbar": {
"position": "fixed"
},
"info": false
});
});
</script>

</body>

</html>
