---
title: ""
always_allow_html: yes
output:
  rmarkdown::html_document:
        toc: false
        toc_depth: 2
        number_sections: false
        theme: readable
        highlight: default
        toc_float: false
        code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(results="asis")
knitr::opts_chunk$set(cache=FALSE)
knitr::opts_chunk$set(echo=FALSE)
knitr::opts_chunk$set(warning=FALSE)
knitr::opts_chunk$set(message=FALSE)
knitr::opts_chunk$set(fig_cap="yes")
```

```{r, eval=FALSE, echo=FALSE}
cd ~/Sites/XGR/XGRplus-site
################################################
R
library(tidyverse)
## rmarkdown::render + bookdown::html_document2
rmarkdown::render("now.RMD_EAsnp.Rmd", bookdown::html_document2(number_sections=F,theme=c("readable","united")[2], hightlight="default"), output_dir="/Users/hfang/Sites/XGR/XGRplus-site/app/examples")

q('no')
################################################
```


# {.tabset .tabset-fade}


## **Input SNP Information**
Completed on `r format(Sys.time(), "%a %b %d %X %Y")` (`r Sys.timezone()`), with the runtime of `r ls_rmd$runtime`.

> **An interactive table** for user-input SNPs (n=`r nrow(ls_rmd$data_input)`, being processed). The table includes three columns: 1st column `SNPs` for input dbSNP rsIDs, 2nd column `P-values` for the significance info, and 3rd column `Locus` for genomic locations (the genome build `hg19`). Each SNP listed in the column `SNPs` is hyperlinked to its corresponding `dbSNP` page.

```{r}
#input.file <- "~/Sites/XGR/XGRplus-site/my_xgrplus/public/app/examples/eg_EAsnp.txt"
#data_input <- read_delim(input.file, delim="\t", col_names=F) %>% as_tibble() %>% set_names("Genes")

data_input <-  ls_rmd$data_input %>% mutate(SNPs=str_c("<a href='https://www.ncbi.nlm.nih.gov/snp/", SNPs, "' target='_blank'>&nbsp;", SNPs, "</a>"))

data_input %>% DT::datatable(options=list(pageLength=5,searchHighlight=T,dom='Bfrtip',buttons=c('csv','copy'), dom='t',scrollX=T), style='default', caption="", rownames=FALSE, escape=F, extensions=c('Buttons'))
```


## **Output: Linked Genes**

> **Linked Gene table** contains information on genes (n=`r ls_rmd$num_LG`, linked from input SNPs), including the column `GScores` (ranged from 1 to 10) that quantifies the degree to which genes are responsible for genetic associations. `Linked genes` are sorted by their linked gene scores and are hyperlinked to [GeneCards](https://www.genecards.org) for further information. Please refer to **`Evidence table`** for details on the linking evidence.

```{r}
#filename <- str_c('/Users/hfang/Sites/XGR/PiER-site/app/examples/', 'eV2CG.xlsx')
filename <- ls_rmd$xlsx_LG

message(sprintf("'%s' '%s' (%s) ...", filename, getwd(), as.character(Sys.time())), appendLF=TRUE)

tb <- openxlsx::read.xlsx(filename) %>% as_tibble() %>% mutate(`Linked genes`=sprintf("<a href='https://www.genecards.org/cgi-bin/carddisp.pl?gene=%s' target='%s''>%s</a>", Gene, Gene, Gene)) %>% transmute(`Linked genes`, GScores=signif(LScore,digits=4), Description)

tb %>% DT::datatable(options=list(pageLength=5,searchHighlight=T,dom='Bfrtip',buttons=c('csv','copy'), dom='t',scrollX=T), style='default', caption="", rownames=FALSE, escape=F, extensions=c('Buttons'))
```

> **Evidence table** displays information on which SNPs (see the column `SNPs`) are used to define linked genes (the column `Linked genes`) based on which evidence (see the column `Evidence`). The column `SNP type` indicates whether the SNP is an input SNP (`Input`) or an additional SNP in linkage disequilibrium (`LD`). The column `Evidence` contains information on the datasets used, such as `Proximity` (indicative of SNPs in proximity), the prefix `PCHiC_` (promoter capture Hi-C datasets), and the prefix `eQTL_` or `pQTL_` (e/pQTL datasets).

```{r}
#filename <- str_c('/Users/hfang/Sites/XGR/PiER-site/app/examples/', 'eV2CG_evidence.xlsx')
filename <- ls_rmd$xlsx_LG_evidence

message(sprintf("'%s' '%s' (%s) ...", filename, getwd(), as.character(Sys.time())), appendLF=TRUE)

tb <- openxlsx::read.xlsx(filename) %>% as_tibble() %>% transmute(SNPs=SNP, `SNP type`=SNP_type, `Linked genes`=Gene, Evidence)

tb %>% DT::datatable(options=list(pageLength=5,searchHighlight=T,dom='Bfrtip',buttons=c('csv','copy'), dom='t',scrollX=T), style='default', caption="", rownames=FALSE, escape=F, extensions=c('Buttons'))
```


## **Output: Enriched Terms**

> **An interactive table** that displays enriched ontology terms based on `r ls_rmd$num_LG` linked genes. Each enriched term (the columns `Ontology`, `Term ID` and `Term Name`) has the enrichment z-score (`Z-score`), the enrichment significance level (`FDR`), odds ratio (`Odds ratio`) and 95% confidence intervals (`95% CI`), the number (`Num`) and list (`Members`) of member genes that overlap with linked genes. The table is sorted by FDR ascendingly.

```{r}
#filename <- str_c('~/Sites/XGR/XGRplus-site/app/examples/', 'dcGO_enrichment.xlsx')
filename <- ls_rmd$xlsx_enrichment

message(sprintf("'%s' '%s' (%s) ...", filename, getwd(), as.character(Sys.time())), appendLF=TRUE)

#tb <- openxlsx::read.xlsx(filename) %>% as_tibble() %>% transmute(Ontology=namespace, `Term ID`=str_c("<a href='http://127.0.0.1:3010/dcGO/", namespace, "/", name, "' target='_blank'>&nbsp;", name, "</a>"), `Term Name`=id, `Z-score`=zscore, FDR=signif(adjp,digits=4), `Odds ratio`=or, `95% CI`=str_c('[',CIl,', ',CIu,']'), Num=nO, Members=overlap) %>% arrange(FDR)

tb <- openxlsx::read.xlsx(filename) %>% as_tibble() %>% transmute(Ontology=namespace, `Term ID`=id, `Term Name`=name, `Z-score`=zscore, FDR=signif(adjp,digits=4), `Odds ratio`=or, `95% CI`=str_c('[',CIl,', ',CIu,']'), Num=nO, Members=overlap) %>% arrange(FDR)

tb %>% DT::datatable(options=list(pageLength=1,searchHighlight=T,dom='Bfrtip',buttons=c('csv','copy'), dom='t',scrollX=T), style='default', caption="", rownames=FALSE, escape=F, extensions=c('Buttons'))
```


## **Output: Dotplot**

> **A dotplot** of enriched ontology terms identified using Fisher's exact test. An editable pdf file is also available for download.

```{r}
#hlink_pdf <- oAuxEmbed("~Sites/XGR/XGRplus-site/app/examples/EAsnp_enrichment_dotplot.pdf", link.text="Download PDF")
hlink_pdf <- oAuxEmbed(ls_rmd$pdf_dotplot, link.text="DOWNLOAD .pdf")
```
<div class="text-right">`r hlink_pdf`</div>


```{r dotplot, fig.cap="A dotplot of enriched ontology terms. The dots colored in dark green represent the terms with FDR < 0.05, indicating statistical significance. The top 5 significant terms (ordered by FDR ascendingly) are labelled for easier interpretation.", fig.width=5, fig.height=4, echo=FALSE, eval=T, fig.align="center",out.width="55%"}
library(png)
library(grid)
#filename <- str_c('~Sites/XGR/XGRplus-site/app/examples/EAsnp_enrichment_dotplot.png')
filename <- ls_rmd$png_dotplot
img <- readPNG(filename)
grid.raster(img)
```


## **Output: Forest Plot**

> **A forest plot** of enriched ontology terms identified using Fisher's exact test. An editable pdf file is also available for download.

```{r}
#hlink_pdf <- oAuxEmbed("~Sites/XGR/XGRplus-site/app/examples/EAsnp_enrichment_forestplot.pdf", link.text="Download PDF")
hlink_pdf <- oAuxEmbed(ls_rmd$pdf_forestplot, link.text="DOWNLOAD .pdf")
```
<div class="text-right">`r hlink_pdf`</div>


```{r forest, fig.cap="Forest plot for enriched ontology terms (ordered by odds ratio descendingly).", fig.width=5, fig.height=4, echo=FALSE, eval=T, fig.align="center",out.width="55%"}
library(png)
library(grid)
#filename <- str_c('"~Sites/XGR/XGRplus-site/app/examples/EAsnp_enrichment_forestplot.png')
filename <- ls_rmd$png_forestplot
img <- readPNG(filename)
grid.raster(img)
```
