---
title: ""
always_allow_html: yes
output:
  rmarkdown::html_document:
        toc: false
        toc_depth: 2
        number_sections: false
        theme: readable
        highlight: default
        toc_float: false
        code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(results="asis")
knitr::opts_chunk$set(cache=FALSE)
knitr::opts_chunk$set(echo=FALSE)
knitr::opts_chunk$set(warning=FALSE)
knitr::opts_chunk$set(message=FALSE)
knitr::opts_chunk$set(fig_cap="yes")
```

```{r, eval=FALSE, echo=FALSE}
cd ~/Sites/XGR/XGRplus-site
################################################
R
library(tidyverse)
## rmarkdown::render + bookdown::html_document2
rmarkdown::render("now.RMD_SAsnp.Rmd", bookdown::html_document2(number_sections=F,theme=c("readable","united")[2], hightlight="default"), output_dir="/Users/hfang/Sites/XGR/XGRplus-site/app/examples")

q('no')
################################################
```


# {.tabset .tabset-fade}


## **Input SNP-Level Summary Data**
Completed on: `r format(Sys.time(), "%a %b %d %X %Y")` (`r Sys.timezone()`), with the runtime of (`r ls_rmd$runtime`).

> **An interactive table** for user-input SNPs (n=`r nrow(ls_rmd$data_input)`, being processed). This table has three columns: 1st column `SNPs` for dbSNP rsIDs (hyperlinked to `dbSNP`), 2nd column `P-values` for use-input significance information, and 3rd column `Locus` for genomic locations (the genome build `hg19`).

```{r}
#input.file <- "~/Sites/XGR/XGRplus-site/my_xgrplus/public/app/examples/eg_Gene.txt"
#data_input <- read_delim(input.file, delim="\t", col_names=F) %>% as_tibble() %>% set_names("Genes")

data_input <-  ls_rmd$data_input %>% mutate(SNPs=str_c("<a href='https://www.ncbi.nlm.nih.gov/snp/", SNPs, "' target='_blank'>&nbsp;", SNPs, "</a>"))

data_input %>% DT::datatable(options=list(pageLength=5,searchHighlight=T,dom='Bfrtip',buttons=c('csv','copy'), dom='t',scrollX=T), style='default', caption="", rownames=FALSE, escape=F, extensions=c('Buttons'))
```


## **Output: Linked Genes**

> **Linked Gene table** contains information on genes (n=`r ls_rmd$num_LG`, linked from the input SNPs), including the column `GScores` (ranged from 1 to 10) that quantifies the degree to which genes are responsible for genetic associations. `Linked genes` are sorted by their linked gene scores and are hyperlinked to [GeneCards](https://www.genecards.org) for further information. Please refer to **`Evidence table`** for details on the linking evidence.

```{r}
#filename <- str_c('/Users/hfang/Sites/XGR/PiER-site/app/examples/', 'eV2CG.xlsx')
filename <- ls_rmd$xlsx_LG

message(sprintf("'%s' '%s' (%s) ...", filename, getwd(), as.character(Sys.time())), appendLF=TRUE)

tb <- openxlsx::read.xlsx(filename) %>% as_tibble() %>% mutate(`Linked genes`=sprintf("<a href='https://www.genecards.org/cgi-bin/carddisp.pl?gene=%s' target='%s''>%s</a>", Gene, Gene, Gene)) %>% transmute(`Linked genes`, GScores=signif(LScore,digits=4), Description)

tb %>% DT::datatable(options=list(pageLength=5,searchHighlight=T,dom='Bfrtip',buttons=c('csv','copy'), dom='t',scrollX=T), style='default', caption="", rownames=FALSE, escape=F, extensions=c('Buttons'))
```

> **Evidence table** displays information on which SNPs (see the column `SNPs`) are used to define linked genes (the column `Linked genes`) based on which evidence (see the column `Evidence`). The column `SNP type` indicates whether the SNP is an input SNP (`Input`) or an additional SNP in linkage disequilibrium (`LD`). The column `Evidence` contains information on the datasets used, such as `Proximity` (indicative of SNPs in proximity), the prefix `PCHiC_` (promoter capture Hi-C datasets), and the prefix `eQTL_` or `pQTL_` (e/pQTL datasets).

```{r}
#filename <- str_c('/Users/hfang/Sites/XGR/PiER-site/app/examples/', 'eV2CG_evidence.xlsx')
filename <- ls_rmd$xlsx_LG_evidence

message(sprintf("'%s' '%s' (%s) ...", filename, getwd(), as.character(Sys.time())), appendLF=TRUE)

tb <- openxlsx::read.xlsx(filename) %>% as_tibble() %>% transmute(SNPs=SNP, `SNP type`=SNP_type, `Linked genes`=Gene, Evidence)

tb %>% DT::datatable(options=list(pageLength=5,searchHighlight=T,dom='Bfrtip',buttons=c('csv','copy'), dom='t',scrollX=T), style='default', caption="", rownames=FALSE, escape=F, extensions=c('Buttons'))
```


## **Output: Gene Subnetwork**

> **A network visualisation** of the subnetwork, with `r ls_rmd$vcount` genes/nodes color-coded by linked gene scores (ranged from 1 to 10).`r if(ls_rmd$combinedP<1){str_c(" The significance (empirical p-value) of observing the identified crosstalk by chance is ",as.character(ls_rmd$combinedP),", as estimated by a degree-preserving node permutation test. ")}` An editable pdf file is also available for download.


```{r}
hlink_pdf <- oAuxEmbed(ls_rmd$pdf_crosstalk, link.text="DOWNLOAD .pdf")
```
<div class="text-right">`r hlink_pdf`</div>

```{r crosstalk, fig.cap="Visualisation of the identified subnetwork, with genes/nodes color-coded by linked gene scores (ranged from 1 to 10).", fig.width=6, fig.height=6, echo=FALSE, eval=T, fig.align="center",out.width="50%"}
library(png)
library(grid)
#filename <- str_c('~Sites/XGR/XGRplus-site/app/examples/SAsnp_crosstalk.png')
filename <- ls_rmd$png_crosstalk
img <- readPNG(filename)
grid.raster(img)
```

<hr>

> **An interactive table** for the subnetwork genes (n=`r ls_rmd$vcount`), with the column `GScores` for linked gene scores (ranged from 1 to 10). Each gene listed in the column `Linked genes` is hyperlinked to its corresponding [GeneCards](https://www.genecards.org) page.

```{r}
#filename <- str_c('/Users/hfang/Sites/XGR/PiER-site/app/examples/', 'dTarget_crosstalk.xlsx')
filename <- ls_rmd$xlsx_crosstalk

message(sprintf("'%s' '%s' (%s) ...", filename, getwd(), as.character(Sys.time())), appendLF=TRUE)

tb <- openxlsx::read.xlsx(filename) %>% as_tibble() %>% mutate(Genes=sprintf("<a href='https://www.genecards.org/cgi-bin/carddisp.pl?gene=%s' target='%s''>%s</a>", Genes, Genes, Genes)) %>% transmute(`Linked genes`=Genes, GScores=signif(-log10(Pvalue),digits=4), Description)

tb %>% DT::datatable(options=list(pageLength=5,searchHighlight=T,dom='Bfrtip',buttons=c('csv','copy'), dom='t',scrollX=T), style='default', caption="", rownames=FALSE, escape=F, extensions=c('Buttons'))
```

